### PR TITLE
Optimize isSharingDisabledForUser

### PR DESCRIPTION
**Pull request by rullzer**
I was doing some benchmarking and profiling for the common propfinds of the client. And on a directory with 20K entries I noticed a significant impact of the isSharingDisabledForUser function.

[Blackfire graph](https://blackfire.io/profiles/compare/939ac718-4d11-4d2a-8c7a-87ce89b6b13c/graph)

Our own perf tool (https://github.com/owncloud/administration/tree/master/performance-tests-c%2B%2B) also show improvements for the propfinds:

|  | master | branch |
| --- | --- | --- |
| RequestEtagJob | 251.1 ms | 234.1 ms |
| ConnectionValidator | 267.1 ms | 256.3 ms |
| DiscoverySingleDirectoryJob | 612.8 ms | 529.0 ms |
| CheckQuotaJob | 253.0 ms | 246.3 ms |

Eventaully we should properly inject the sharemanager. But this was already such a nice preformance improvement.

CC: @icewind1991 @PVince81 @MorrisJobke @nickvergessen

### CODE REVIEW

File: lib/public/util.php [hunk 1]
```diff
@@ -171,13 +171,16 @@ public static function logException( $app, \Exception $ex, $level = \OCP\Util::F
...
 	 * @since 7.0.0
+	 * @deprecated 9.1.0 Use \OC::$server->getShareManager()->sharingDisabledForUser
 	 */
 	public static function isSharingDisabledForUser() {
-		return \OC_Util::isSharingDisabledForUser(
-				\OC::$server->getConfig(),
-				\OC::$server->getGroupManager(),
-				\OC::$server->getUserSession()->getUser()
-		);
+		static $shareManager;
```
Comment (PVince81) [inline lib/public/util.php]: boo, define this as attribute. Please no local static!

Comment (nickvergessen) [inline lib/public/util.php]: yeah, well attribute doesnt help on a static class

**Comment by PVince81** [comment]
Wow, okay. Didn't think even that this part would cause issues

File: lib/public/util.php (hunk 1 above)
Comment (PVince81) [inline lib/public/util.php]: it does make it more visible so we can see it globally.

Comment (rullzer) [inline lib/public/util.php]: fair enough... altough private static members are also nasty

**Comment by rullzer** [comment]
@PVince81 well 20K entries is a lot... but the saving is also significant...

**Comment by rullzer** [comment]
Aaah mmm it seems I broke a few tests... let me fix that tomorrow... I think it is because it are semi intergration tests so we cache the result but don't reset the cache..

**Comment by PVince81** [comment]
@rullzer please rebase

**Comment by rullzer** [comment]
Yeah those tests don't play nice... I'll set this back to to develop.. need to look deeper... or just fix it properly at some point...

**Comment by rullzer** [comment]
OK unit tests pass now at least locally...
review time!

File: .htaccess [hunk 2]
```diff
@@ -72,3 +72,7 @@ Options -Indexes
 <IfModule pagespeed_module>
   ModPagespeed Off
 </IfModule>
+#### DO NOT CHANGE ANYTHING ABOVE THIS LINE ####
+
+ErrorDocument 403 /core/templates/403.php
+ErrorDocument 404 /core/templates/404.php
```
Comment (icewind1991) [inline .htaccess]: accidental added file

**Comment by icewind1991** [comment]
:+1: looks good otherwise

File: .htaccess (hunk 2 above)
Comment (rullzer) [inline .htaccess]: Ah yes the `-a` ;)

Comment (rullzer) [inline .htaccess]: Fixed

**Comment by PVince81** [comment]
Seems to work fine :+1:

**Pull request merged**